### PR TITLE
#2488 : API for in/out download from a list of processingIds

### PR DESCRIPTION
**Pull request by DuckflipXYZ**
#2488 : API for in/out download from a list of processingIds

This PR  is only relative to the creation of an API, there isn't any Shanoir interface yet, and it's currently specific to CometeMoelle project. Please be aware of the [fli-iam](https://github.com/fli-iam/py_noir/issues/6) feature, which implements a python script for generating API calls to that new API

(Tested in local and in qualif)

### CODE REVIEW

**Comment by DuckflipXYZ** [comment]
Some tests have been modified for allowing "extends" implementation. When a class is extended by an other one, all its @Autowired variable class declaration need to get @Qualifier, for knowing which bean load among the extended class and the extension class beans (usefull when there are multiple extensions class of the same extended class, those extensions class are not implementing the same methods ...)

**Comment by DuckflipXYZ** [comment]
Extraction can now be realized with examination ids instead of processing ids. Refer to the https://github.com/fli-iam/py_noir/issues/6 feature.

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/processing/controler/DatasetProcessingApi.java [hunk 1]
```diff
@@ -120,4 +122,34 @@ ResponseEntity<Void> updateDatasetProcessing(
...
+			@ApiResponse(responseCode = "200", description = "zip file"),
+			@ApiResponse(responseCode = "401", description = "unauthorized"),
+			@ApiResponse(responseCode = "403", description = "forbidden"),
+			@ApiResponse(responseCode = "404", description = "no dataset found"),
+			@ApiResponse(responseCode = "500", description = "unexpected error") })
+	@GetMapping(value = "/massiveDownloadByProcessing")
+	@PreAuthorize("hasRole('ADMIN') or (hasAnyRole('EXPERT', 'USER') and @datasetSecurityService.HasRightOnEveryDatasetOfProcessings(#processingIds, 'CAN_DOWNLOAD'))")
+	void massiveDownloadByProcessingId(
+			@Parameter(description = "id of the processing", required=true) @Valid
+			@RequestParam(value = "processingIds") List<Long> processingIds,
```
Comment (michaelkain) [inline shanoir-ng-datasets/src/main/java/org/shanoir/ng/processing/controler/DatasetProcessingApi.java:136]: Hi @DuckflipXYZ, thank you for this good PR. I am not sure, if sending the list of processingIds as request param is a good idea, as limited by number of chars, a Post would be a better way to do so? same for the method below

Comment (michaelkain) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/processing/controler/DatasetProcessingApi.java:136]: and please harmonize the method names in GetMapping with @Operation above

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/processing/controler/DatasetProcessingApi.java [hunk 2]
```diff
@@ -120,4 +122,34 @@ ResponseEntity<Void> updateDatasetProcessing(
...
+	@Operation(summary = "massiveDownloadProcessingDatasetsByExaminationIds", description = "If exists, returns a zip file of the inputs/outputs per processing corresponding to the given examination IDs")
+	@ApiResponses(value = {
+			@ApiResponse(responseCode = "200", description = "zip file"),
+			@ApiResponse(responseCode = "401", description = "unauthorized"),
+			@ApiResponse(responseCode = "403", description = "forbidden"),
+			@ApiResponse(responseCode = "404", description = "no dataset found"),
+			@ApiResponse(responseCode = "500", description = "unexpected error") })
+	@GetMapping(value = "/massiveDownloadProcessingByExamination")
+	@PreAuthorize("hasRole('ADMIN') or (hasAnyRole('EXPERT', 'USER') and @datasetSecurityService.hasRightOnExaminations(#examinationIds, 'CAN_DOWNLOAD'))")
+	void massiveDownloadProcessingByExaminationId(
```
Comment (michaelkain) [inline shanoir-ng-datasets/src/main/java/org/shanoir/ng/processing/controler/DatasetProcessingApi.java:149]: same here: 3 names for the same method, in @Operation, @GetMapping and method name, please harmonize to something unique and precise

Comment (michaelkain) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/processing/controler/DatasetProcessingApi.java:149]: Could you document, what is the expected output of both calls, how does it look like on the disk, once downloaded

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/processing/controler/DatasetProcessingApiController.java [hunk 3]
```diff
@@ -147,12 +164,72 @@ public ResponseEntity<Void> updateDatasetProcessing(
...
+					throw new Exception();
+				}
+				processing = datasetProcessingService.findById(processingId).get();
+				processingList.add(processing);
+			}catch (Exception e) {
+				throw new RestServiceException(
+						new ErrorModel(HttpStatus.FORBIDDEN.value(), processingId + " is not a valid processing id."));
+			}
+
+			if(!Objects.equals(processing.getComment(), "comete_moelle/0.1")){
```
Comment (michaelkain) [inline shanoir-ng-datasets/src/main/java/org/shanoir/ng/processing/controler/DatasetProcessingApiController.java:198]: I would add a filter param here, we can not keep this in the general code, so the caller decides, what to filter out

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/processing/controler/DatasetProcessingApiController.java [hunk 4]
```diff
@@ -147,12 +164,72 @@ public ResponseEntity<Void> updateDatasetProcessing(
...
+				if(Objects.isNull(examination)){
+					throw new Exception();
+				}
+				examinationList.add(examination);
+			}catch (Exception e) {
+				throw new RestServiceException(
+						new ErrorModel(HttpStatus.FORBIDDEN.value(), examinationId + " is not a valid examination id."));
+			}
+		}
+		processingDownloaderService.massiveDownloadByExamination(examinationList, resultOnly, "dcm" , response, false, null);
```
Comment (michaelkain) [inline shanoir-ng-datasets/src/main/java/org/shanoir/ng/processing/controler/DatasetProcessingApiController.java:233]: ByExaminations, as multiple are used

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/processing/service/ProcessingDownloaderServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+        if (!downloadResults.isEmpty()) {
+            ZipEntry zipEntry = new ZipEntry(JSON_RESULT_FILENAME);
+            zipEntry.setTime(System.currentTimeMillis());
+            zipOutputStream.putNextEntry(zipEntry);
+            zipOutputStream.write(objectMapper.writeValueAsString(downloadResults).getBytes());
+            zipOutputStream.closeEntry();
+        }
+    }
+
+    private void manageDatasetDownload(Dataset dataset, Map<Long, DatasetDownloadError> downloadResults, ZipOutputStream zipOutputStream, String subjectName, String processingFilePath, String format, boolean withManifest, Map<Long, List<String>> filesByAcquisitionId, Long converterId) throws IOException, RestServiceException {
```
Comment (michaelkain) [inline shanoir-ng-datasets/src/main/java/org/shanoir/ng/processing/service/ProcessingDownloaderServiceImpl.java:98]: is this method different to the existing already?

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/security/DatasetSecurityService.java [hunk 6]
```diff
@@ -1103,5 +1103,60 @@ public boolean hasRightOnExamination(String examinationUID, String rightStr) thr
...
+	public boolean HasRightOnEveryDatasetOfProcessings(List<Long> processingIds, String rightStr) {
+		boolean hasRight = true;
+
+		for (Long processingId : processingIds) {
+			if (KeycloakUtil.getTokenRoles().contains("ROLE_ADMIN") || processingId == null) {
+				continue;
+			}
+			Iterable<Dataset> datasets = datasetRepository.findDatasetsByProcessingId(processingId);
+
+			for (Dataset dataset : datasets) {
```
Comment (michaelkain) [inline shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/security/DatasetSecurityService.java:1116]: I wonder, if the below code exists already on the dataset level?

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/processing/controler/DatasetProcessingApi.java (hunk 2 above)
Comment (DuckflipXYZ) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/processing/controler/DatasetProcessingApi.java]: I added to the description of the @Operation : "Datas are in the http response body, it must be written in a zip file. Datas are sorted with folders according to their respective examination and processing."

It seems enough for me, tell me if it's enough for you

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/processing/service/ProcessingDownloaderServiceImpl.java (hunk 5 above)
Comment (DuckflipXYZ) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/processing/service/ProcessingDownloaderServiceImpl.java]: Yes, the current code division in DatasetDownloaderServiceImpl does not allow to use this specific part of code. I would not change the existing code, that's why I did that way.

I did a specific commit for a code division update enabling to use that part of code in the children class, as I'm supposing you wanted it to be like that. Feel free to remove it or not at PR merging, it won't break the functionality

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/security/DatasetSecurityService.java (hunk 6 above)
Comment (DuckflipXYZ) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/security/DatasetSecurityService.java]: Same as above

**Comment by DuckflipXYZ** [comment]
When  merging this PR, please check the potential merge of those 2 associated py_noir PR :
- https://github.com/fli-iam/py_noir/pull/9
- https://github.com/fli-iam/py_noir/pull/8

**Review by michaelkain — APPROVED** [review]
Dear Guewen,
really good job! Thank you a lot!
Now it looks fine for me. I approve by code review only. I did not test locally
as I do not take the time to set up everything. It would be great, if you could
make me a demo on Friday morning, using qualif please. Thank you, Michael

**Pull request merged**